libcasr/src/stacktrace.rs

```diff
@@ -287,11 +288,226 @@ impl Filter for Stacktrace {
             self.drain(pos + 1..);
         }
 
-        // Remove trusted functions from stack trace
+        // Find repeating intervals in stacktrace
+        let mut indices = Vec::new();
+        indices.resize(self.len(), true);
+        // We need to create an element that will not be equal to others
+        let mut not_an_element = StacktraceEntry::default();
+        not_an_element.debug.file = "not_equal_to_other_names".to_string();
+        not_an_element.debug.line = 1;
+        if let Ok(intervals) = get_interval_repetitions(self, &not_an_element) {
+            intervals.iter().for_each(|(start, end, seq_len)| {
+                (start + seq_len..end + 1).for_each(|idx| indices[idx] = false)
+            });
+        }
+        let mut keep = indices.iter();
+
+        // Remove repeating intervals and trusted functions from stack trace
         self.retain(|entry| {
-            (entry.function.is_empty() || !rfunction.is_match(&entry.function))
+            *keep.next().unwrap()
+                && (entry.function.is_empty() || !rfunction.is_match(&entry.function))
                 && (entry.module.is_empty() || !rfile.is_match(&entry.module))
                 && (entry.debug.file.is_empty() || !rfile.is_match(&entry.debug.file))
         });
     }
 }
+
+/// Map indices i of arr to the length of the longest border of arr[1 .. i]
+///
+/// # Arguments
+///
+/// * `arr` - given sequence
+///
+/// # Return value
+///
+/// Vector of mapped indices
+fn prefix_function<T: PartialEq>(arr: &[T]) -> Vec<usize> {
+    let mut z = Vec::new();
+    let (n, mut r, mut l) = (arr.len(), 0, 0);
+    z.resize(n, 0);
+    for i in 1..n {
+        if i <= r {
+            z[i] = z[i - l].min(r - i + 1);
+        }
+        while i + z[i] < n && arr[z[i]] == arr[i + z[i]] {
+            z[i].add_assign(1);
+        }
+        if i + z[i] - 1 > r {
+            l = i;
+            r = i + z[i] - 1;
+        }
+    }
+    z
+}
+
+/// Add interval to the resulting vector
+///
+/// # Arguments
+///
+/// * `intervals` - resulting vector
+///
+/// * `shift` - shift
+///
+/// And other algorithm parameters.
+/// More about Main-Lorentz algorithm https://cp-algorithms.com/string/main_lorentz.html
+fn add_interval(
+    intervals: &mut Vec<(usize, usize, usize)>,
+    shift: usize,
+    left: bool,
+    cntr: usize,
+    l: usize,
+    k1: usize,
+    _k2: usize,
+) {
+    let l1 = k1.min(l);
```

Comment:
int l1 = max(1, l - k2)

Reply:
if k2=2, l=2, k1=0 algorithm will not work correctly. In cp-algorithm, they just exit the loop in this case, but this is a valid situation.